crates/wysiwyg/src/dom/insert_parent.rs

```diff
@@ -0,0 +1,271 @@
+
+use super::{Dom, Range};
+
+impl<S> Dom<S>
+where
+    S: UnicodeString,
+{
+    #[allow(dead_code)]
+    /// Insert a node and make this node the parent of a given range.
+    pub fn insert_parent(&mut self, range: &Range, mut new_node: DomNode<S>) {
+        #[cfg(any(test, feature = "assert-invariants"))]
+        self.assert_invariants();
+
+        if range.is_empty() {
+            panic!("Cannot add a parent to an empty range");
+        }
+
+        if !matches!(new_node, DomNode::Container(_)) {
+            panic!("New node is not a container");
+        }
+        // Check if the range has shared ancestry. The new parent node should not contain
+        // these nodes, so filter them from the range.
+        let shared_depth = range.shared_parent().depth();
+        let range = Range::new(range.locations_from_depth(shared_depth));
+
+        // Prepare the new parent node to have the selected range moved into it:
+        // - Set the new node's handle to the start of the range
+        // - Get the mutable container node to move the selected range into
+        let new_handle: DomHandle = range.node_handles().min().unwrap().clone();
+        let new_handle = if range.leaves().min().unwrap().is_covered() {
+            new_handle
+        } else {
+            // If the first leaf in the range is completely covered, then leave space
+            // for the first part of the existing split leaf.
+            new_handle.next_sibling()
+        };
+        new_node.set_handle(new_handle.clone());
+        let container = new_node.as_container_mut().unwrap();
+
+        // Remove the selected nodes from the DOM and add them to the new container
+        for location in range.locations.iter().rev() {
+            let node = &location.node_handle;
+
+            // If the location is covered, it can be moved to the container.
+            if location.is_covered() {
+                // Ignore children of covered nodes which are moved in their parents.
+                if node.depth() != shared_depth + 1 {
+                    continue;
+                }
+                let node = self.remove(node);
+                container.insert_child(0, node);
+            }
+            // ... else handle partially covered nodes
+            else {
+                // If the a leaf is partially included at the start of the range, split the tree
+                // and add the end part to the new container.
+                // If the leaf is partially included at the end of the range, split the tree and
+                // add the start part to the new container.
+                if !self.lookup_node(node).is_leaf() {
+                    continue;
+                }
+
+                let offset = if location.ends_inside() {
+                    location.end_offset
+                } else {
+                    location.start_offset
+                };
+
+                let (left, left_handle, right, right_handle) =
```

Comment:
This is actually what I was a bit afraid of. If I'm not mistaken, for both the start and end leaves, as well as their direct ancestors (since you're using `range.locations` instead of `range.leaves()`), you're duplicating the whole Dom, then splitting it.

Reply:
We are iterating over `range.locations` but this code is only run on the leaves due to [lines 60-62](https://github.com/matrix-org/matrix-rich-text-editor/pull/415/files#diff-cfbcd9fb6c294410b18d63f5d384054ff8f86b4a956e5fe98518a346f3ea5861R60-R62):
```
if !self.lookup_node(node).is_leaf() {
    continue;
}
```

Reply:
Oh, I see. Sorry, I completely missed those lines 👍 .